### PR TITLE
Improve Line Drawing performance

### PR DESCRIPTION
**Pull request by SZenglein**
Improve Line drawing performance by taking a smaller context from the
canvas.

### CODE REVIEW

**Comment by SZenglein** [comment]
resolves #88 

Note that this does not change much of the current implementation. It just takes a smaller section of the canvas to draw on.

**Comment by Bios-Marcel** [comment]
I'll be waiting with this PR for one or two days. As I am doing a huge refactoring right now. Don't worry though, I'll resolve conflicts later on :)

File: templates/lobby.html [hunk 1]
```diff
@@ -649,22 +649,30 @@
...
-        const imageData = context.getImageData(0, 0, context.canvas.width, context.canvas.height);
 
+        let radius = Math.floor(lineWidth / 2);
+        // calculate bounding box
+        let left = Math.max(0, Math.min(x1, x2) - lineWidth);
+        let top = Math.max(0, Math.min(y1, y2) - lineWidth);
+        let right = Math.max(0, Math.min(context.canvas.width, Math.max(x1, x2) + lineWidth));
+        let bottom = Math.max(0, Math.min(context.canvas.height, Math.max(y1, y2) + lineWidth));
+
+        const imageData = context.getImageData(left, top, right - left, bottom - top);
```
Comment (Bios-Marcel) [inline templates/lobby.html:660]: Throws when hitting F5: TypeError: Argument 1 of CanvasRenderingContext2D.getImageData is not a finite floating-point value.

**Comment by SZenglein** [comment]
Strange, that does not make any sense.   Maybe it would need to be rounded again because of some dumb floating point arithmetic.

Unfortunately I don't have this error. What browser are you testing on? 

I'd really like to fix this and get it merged.

**Comment by Bios-Marcel** [comment]
Firefox. Anyway,the whole code has become quite confusing. Scaling, DPI and now zoom also seems to cause problems. I might just do another refactor :D

**Comment by SZenglein** [comment]
Oof. I mean, I just made the imagedata a bit smaller, should not break much.

**Comment by Bios-Marcel** [comment]
Well, apparently :man_shrugging: 

I also didn't write the line algorithm myself. It's not the default bresenham. I will probably have to do a big refacotring of all the drawing stuff.

**Comment by SZenglein** [comment]
The line drawing actually works quite well when taking a smaller canvas section. Even mobile devices should not have a problem with it.

**Comment by SZenglein** [comment]
Could you please check out this branch? There was still an error with the bounds checking and it will no longer take an image of height or width zero, which is not allowed. I can not in any way provoke an error now.

https://github.com/SZenglein/scribble.rs/tree/draw_perf_again

**Pull request merged**